File: prime-numbers-finder.py
def prime_sieve(limit):
  """
  Finds all prime numbers up to a given limit using the Sieve of Eratosthenes.

  Args:
    limit: The upper bound for finding prime numbers.

  Returns:
    A list of prime numbers up to the given limit.
  """

  prime = [True for i in range(limit + 1)]
  p = 2
  while (p * p <= limit):
    if (prime[p] == True):
      for i in range(p * p, limit + 1, p):
        prime[i] = False
    p += 1
  prime[0] = False
  prime[1] = False
  primes = []
  for p in range(limit + 1):
    if prime[p]:
      primes.append(p)
  return primes

# Example usage:
limit = 1000000  # Find primes up to 1 million
primes = prime_sieve(limit)
print(primes)

File: prime-numbers-finder.rb
def sieve_of_eratosthenes(n)
  primes = Array.new(n + 1, true)
  primes[0] = primes[1] = false

  (2..Math.sqrt(n).ceil).each do |p|
    next unless primes[p]
    (p * p..n + 1).step(p) { |i| primes[i] = false }
  end

  primes.each_index.select { |i| primes[i] }
end

# Example usage:
limit = 100000
start_time = Time.now
prime_numbers = sieve_of_eratosthenes(limit)
end_time = Time.now

puts "Found #{prime_numbers.length} prime numbers up to #{limit} in #{end_time - start_time} seconds."
# Print the first 20 primes for demonstration
puts prime_numbers[0..19]
